File: l33tcode/substring-with-concatenation-of-all-words.py
from typing import List, Set, Dict
from collections import defaultdict, Counter


class RollingHash:
    def __init__(self) -> None:
        self._hash = 0
        self._length = 0
        self._base = ord("z") - ord("a") + 1

    def add(self, char: str) -> None:
        self._hash *= self._base
        self._hash += ord(char) - ord("a")
        self._length += 1

    def popleft(self) -> None:
        if self._length == 0:
            raise RuntimeError("Nothing to remove")

        self._length -= 1
        self._hash %= self._base ** self._length

    @property
    def value(self) -> int:
        return self._hash


class Solution:
    def findSubstringBruteForce(self, string: str, words: List[str]) -> List[int]:
        matches = defaultdict(list)

        dp: List[List[int]] = [[] for _ in range(len(string))]

        for word in range(len(words)):
            for pos in range(len(string)):
                if string[pos : pos + len(words[word])] == words[word]:
                    matches[words[word]].append(pos)
                    dp[pos].append(word)

        def dfs(pos: int, seen: Set[int]) -> bool:
            if len(seen) == len(words):
                return True

            if pos == len(string):
                return False

            for word in dp[pos]:
                if word not in seen:
                    seen.add(word)
                    if dfs(pos + len(words[word]), seen):
                        return True
                    seen.remove(word)

            return False

        result = []

        for pos in range(len(string)):
            if dfs(pos, set()):
                result.append(pos)

        return result

    def findSubstring(self, string: str, words: List[str]) -> List[int]:
        hashes = [0] * len(string)
        word_len = len(words[0]) if words else 0
        rolling_hash = RollingHash()

        for pos in range(len(string) + 1):
            if pos >= word_len:
                hashes[pos - word_len] = rolling_hash.value

            if pos < len(string):
                rolling_hash.add(string[pos])

            if pos >= word_len:
                rolling_hash.popleft()

        hash_counter: Dict[int, int] = Counter()

        for word in words:
            word_rolling_hash = RollingHash()
            for char in word:
                word_rolling_hash.add(char)
            hash_counter[word_rolling_hash.value] += 1

        total = len(words)

        def dfs(pos: int) -> bool:
            nonlocal total
            if total == 0:
                return True

            if pos >= len(string):
                return False

            if hash_counter[hashes[pos]] > 0:
                hash_counter[hashes[pos]] -= 1
                total -= 1
                if dfs(pos + word_len):
                    hash_counter[hashes[pos]] += 1
                    total += 1
                    return True
                hash_counter[hashes[pos]] += 1
                total += 1

            return False

        result = []

        for pos in range(len(string) - word_len * len(words) + 1):
            if dfs(pos):
                result.append(pos)

        return result

File: l33tcode_rs/src/substring_with_concatenation_of_all_words.rs
#[allow(dead_code)]
struct Solution;

use std::collections::HashMap;

fn dfs(
    string: &String,
    hash_counter: &mut HashMap<u64, u64>,
    hashes: &mut Vec<u64>,
    word_len: usize,
    pos: usize,
    total: &mut usize,
) -> bool {
    if *total == 0 {
        return true;
    }

    if pos >= string.len() {
        return false;
    }

    if *hash_counter.get(&hashes[pos]).unwrap_or(&0) > 0 {
        hash_counter.entry(hashes[pos]).and_modify(|e| *e -= 1);
        *total -= 1;

        if dfs(
            string,
            hash_counter,
            hashes,
            word_len,
            pos + word_len,
            total,
        ) {
            hash_counter.entry(hashes[pos]).and_modify(|e| *e += 1);
            *total += 1;
            return true;
        }
        hash_counter.entry(hashes[pos]).and_modify(|e| *e += 1);
        *total += 1;
    }

    false
}

impl Solution {
    #[allow(dead_code, clippy::needless_pass_by_value)]
    pub fn find_substring(s: String, words: Vec<String>) -> Vec<i32> {
        // TODO: some of the test cases are still failing
        // Identical python solution works, but this is probably due to the fact that python has
        // unbound integers
        if words.is_empty() {
            return vec![];
        }
        let word_len = words[0].len();
        let mut hashes = vec![0; s.len()];

        let chars_to_num: Vec<u64> = s.chars().map(|c| (c as u64) - ('a' as u64)).collect();

        let mut hash = 0_u64;

        for pos in 0..=s.len() {
            if pos >= word_len {
                hashes[pos - word_len] = hash;
            }

            if pos < s.len() {
                hash = hash.wrapping_mul(26).wrapping_add(chars_to_num[pos]);
            }

            if pos >= word_len {
                hash %= 26_u64.saturating_pow(word_len.try_into().unwrap());
            }
        }

        let mut hash_counter = HashMap::new();

        for word in &words {
            let mut hash = 0;

            for c in word.chars() {
                hash = hash * 26 + (c as u64) - ('a' as u64);
            }
            hash_counter
                .entry(hash)
                .and_modify(|e| *e += 1)
                .or_insert(1);
        }

        let mut result = Vec::new();

        #[allow(clippy::range_plus_one)]
        for pos in 0..(s.len() - word_len * words.len() + 1) {
            let mut total = words.len();
            if dfs(
                &s,
                &mut hash_counter,
                &mut hashes,
                word_len,
                pos,
                &mut total,
            ) {
                result.push(i32::try_from(pos).unwrap());
            }
        }

        result
    }
}
